compiler.ts

```diff
@@ -305,15 +322,106 @@ function codeGenExpr(expr: Expr<Type>, env: GlobalEnv): Array<string> {
+        var brStmts = [];
+        //First check whether index is out of bounds, if so, throw error by going to print_str. Then index the string.
+        brStmts.push(
+          ...[
+            `${brObjStmts.join("\n")}`, //Load the string object to be indexed
+            `(local.set $$string_index)`,
+            `${brKeyStmts.join("\n")}`, //Add the index * 4 value to the address
+            `(local.get $$string_index)(i32.load)(i32.gt_s)`,
+            `(if (then (i32.const -1)(call $print_str)(drop)))`, //Check if string index is out of bounds
+            `(local.get $$string_index)`,
+            `(i32.add (i32.mul (i32.const 4)${brKeyStmts.join("\n")}))`, //Add the index * 4 value to the address
+            `(i32.add (i32.const 4))`, //Adding 4 since string length is at first index
+            `(i32.load)`, //Load the ASCII value of the string index
+            `(local.set $$string_val)`, //store value in temp variable
+            `(i32.load (i32.const 0))`, //load value at 0
+            `(i32.const 0)`, //Length of string is 1, but store as 4 for easier checking
+            `(i32.store)`, //Store length of string in the first position
+            `(i32.load (i32.const 0))`, //Load latest free memory
+            `(i32.add (i32.const 4))`, //Add 4 since we have stored string length at beginning
+            `(local.get $$string_val)`, //load value in temp variable
+            "(i32.store)", //Store the ASCII value in the new address
+          ]
+        );
+        //At end of string, we store ASCII value 0 which represents null
+        brStmts.push(
+          ...[
+            `(i32.load (i32.const 0))`, // Load the dynamic heap head offset
+            `(i32.add (i32.const 8))`, // Calc string index offset from heap offset
+            `(i32.const 0)`, // Store ASCII value for 0 (end of string)
+            "(i32.store)", // Store the ASCII value 0 in the new address
+          ]
+        );
+        brStmts.push(
+          ...[
+            "(i32.load (i32.const 0))", // Get address for the indexed character of the string
+            "(i32.const 0)", // Address for our upcoming store instruction
+            "(i32.load (i32.const 0))", // Load the dynamic heap head offset
+            `(i32.add (i32.const 12))`, // Move heap head beyond the string length
+            "(i32.store)", // Save the new heap offset
+          ]
+        );
+        return brStmts;
+      }
+      break;
     default:
       unhandledTag(expr);
   }
 }
 
+function allocateStringMemory(string_val: string): Array<string> {
+  const stmts = [];
+  var i = 1;
+  //Storing the length of the string at the beginning
+  stmts.push(
+    ...[
+      `(i32.load (i32.const 0))`, // Load the dynamic heap head offset
+      `(i32.const ${string_val.length - 1})`, // Store ASCII value for 0 (end of string)
+      "(i32.store)", // Store the ASCII value 0 in the new address
+    ]
+  );
+  while (i != string_val.length + 1) {
```

Comment:
This makes sense, but I wonder if there is some better way to set an array of memory in WASM rather than 1 byte at a time.